src/tools/rbd_mirror/ImageSync.cc

```diff
@@ -269,89 +274,136 @@ void ImageSync<I>::handle_copy_image(int r) {
-           << "snap_name=" << sync_point.snap_name << dendl;
+  m_update_sync_ctx = nullptr;
 
-  Context *finish_op_ctx = nullptr;
-  if (m_local_image_ctx->exclusive_lock != nullptr) {
-    finish_op_ctx = m_local_image_ctx->exclusive_lock->start_op();
+  if (m_canceled) {
+    return;
   }
-  if (finish_op_ctx == nullptr) {
-    derr << ": lost exclusive lock" << dendl;
-    m_local_image_ctx->snap_lock.put_read();
-    m_local_image_ctx->owner_lock.put_read();
-    finish(-EROFS);
+
+  auto sync_point = &m_client_meta->sync_points.front();
+
+  if (m_client_meta->sync_object_count == m_image_copy_object_count &&
+      sync_point->object_number &&
+      (m_image_copy_object_no - 1) == sync_point->object_number.get()) {
+    // update sync point did not progress since last sync
     return;
   }
 
-  // rollback the object map (copy snapshot object map to HEAD)
-  RWLock::WLocker object_map_locker(m_local_image_ctx->object_map_lock);
-  auto ctx = new FunctionContext([this, finish_op_ctx](int r) {
-      handle_copy_object_map(r);
-      finish_op_ctx->complete(0);
-    });
-  m_local_image_ctx->object_map->rollback(snap_id, ctx);
-  m_local_image_ctx->snap_lock.put_read();
-  m_local_image_ctx->owner_lock.put_read();
+  m_updating_sync_point = true;
+
+  m_client_meta_copy = *m_client_meta;
+  m_client_meta->sync_object_count = m_image_copy_object_count;
+  if (m_image_copy_object_no > 0) {
+    sync_point->object_number = m_image_copy_object_no - 1;
+  }
+
+  CephContext *cct = m_local_image_ctx->cct;
+  ldout(cct, 20) << ": sync_point=" << *sync_point << dendl;
+
+  bufferlist client_data_bl;
+  librbd::journal::ClientData client_data(*m_client_meta);
+  ::encode(client_data, client_data_bl);
+
+  Context *ctx = create_context_callback<
+    ImageSync<I>, &ImageSync<I>::handle_update_sync_point>(
+      this);
+  m_journaler->update_client(client_data_bl, ctx);
 }
 
 template <typename I>
-void ImageSync<I>::handle_copy_object_map(int r) {
-  dout(20) << dendl;
+void ImageSync<I>::handle_update_sync_point(int r) {
+  CephContext *cct = m_local_image_ctx->cct;
+  ldout(cct, 20) << ": r=" << r << dendl;
 
-  assert(r == 0);
-  send_refresh_object_map();
+  if (r < 0) {
+    *m_client_meta = m_client_meta_copy;
+    lderr(cct) << ": failed to update client data: " << cpp_strerror(r)
+               << dendl;
+  }
+
+  {
+    Mutex::Locker timer_locker(*m_timer_lock);
+    Mutex::Locker locker(m_lock);
+    m_updating_sync_point = false;
+
+    if (m_image_copy_request != nullptr) {
+      m_update_sync_ctx = new FunctionContext(
+        [this](int r) {
+          this->send_update_sync_point();
+        });
+      m_timer->add_event_after(m_update_sync_point_interval,
+                               m_update_sync_ctx);
+      return;
+    }
+  }
+
+  send_flush_sync_point();
 }
 
 template <typename I>
-void ImageSync<I>::send_refresh_object_map() {
-  dout(20) << dendl;
+void ImageSync<I>::send_flush_sync_point() {
```

Comment:
Nit: is this state needed? Is it just to save up to one throttle period of work should the copy fail / be canceled?

Reply:
I think yes. But we have this state in the current `rbd_mirror::image_sync::ImageCopyRequest` [1] and it looked safer to me just to copy this. If you think it is not needed I will remove it.

[1] https://github.com/ceph/ceph/blob/master/src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc#L319

Reply:
No worries -- just wondering and I didn't realize it was already in the rbd-mirror state machines.

Reply:
Ok, I have returned back `send_flush_sync_point` in my latest push. I think it is ready for review.

BTW, in the current `rbd_mirror::image_sync::ImageCopyRequest::send_flush_sync_point` we are sending `m_client_meta_copy` [1], while I think it should be `m_client_meta`.

[1] https://github.com/ceph/ceph/blob/master/src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc#L337

Reply:
I thought the state machines followed the general idea that they update the in-memory copy in the "send" method and then update ```m_client_meta``` after the update is successful.